server/commands/write.go

```diff
@@ -87,6 +90,63 @@ func (c *WriteCommand) validateTuplesets(ctx context.Context, req *openfgapb.Wri
 	return nil
 }
 
+// validateTypes makes sure that when writing a tuple, the types are compatible.
+// 1. If the tuple is of the form (person:bob, reader, doc:budget), then the type "doc", relation "reader" allows type "person".
+// 2. If the tuple is of the form (group:abc#member, reader, doc:budget), then the type "doc", relation "reader" must allow type "group", relation "member".
+// 3. If the tuple is of type (*, reader, doc:budget), we allow it only if the type "doc" relation "reader" has "allow public" = true
+func (c *WriteCommand) validateTypes(ctx context.Context, store string, authorizationModelID string, tk *openfgapb.TupleKey, dbCallsCounter utils.DBCallCounter) error {
+	objectType, _ := tupleUtils.SplitObject(tk.GetObject()) // e.g. "doc"
+
+	dbCallsCounter.AddReadCall()
+	typeDefinition, err := c.datastore.ReadTypeDefinition(ctx, store, authorizationModelID, objectType)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			return &tupleUtils.TypeNotFoundError{TypeName: objectType}
+		}
+		return err
+	}
+	if typeDefinition.Metadata == nil {
+		// authorization model is old and does not have type information
+		return nil
+	}
+	relationInformation := typeDefinition.Metadata.Relations[tk.Relation]
```

Comment:
I am not sure, but I think you may have to worry if relations is nil? Also if the map contains `tk.Relation`. Perhaps something like 
```
relationInformation, ok := typeDefinition.GetMetadata().GetRelations()[tk.Relation]
```
may help. The `Gets` handle the nils.

Reply:
> The Gets handle the nils.

TIL! Thank you.